src/dataflow/src/render/sources.rs

```diff
@@ -816,6 +862,148 @@ where
+    }
+
+    /// Enables compaction of sources associated with the identifier.
+    pub fn allow_compaction(&mut self, id: GlobalId, frontier: Antichain<Timestamp>) {
+        if let Some(compaction_handle) = self.compaction_handles.get(&id) {
+            let compaction_handle = match compaction_handle.upgrade() {
+                Some(handle) => handle,
+                None => {
+                    // Clean up the expired weak handle.
+                    // TODO: A better (long term) solution might be the addition of a
+                    // command that tells us to drop a source. Right now, sources are
+                    // being droppped implicitly when all their consumers are being
+                    // dropped.
+                    self.compaction_handles.remove(&id);
+                    return;
+                }
+            };
+
+            compaction_handle.allow_compaction(frontier);
+        }
+    }
+
+    /// Removes the maintained state for source `id`.
+    pub fn del_source(&mut self, id: &GlobalId) -> bool {
+        self.compaction_handles.remove(&id).is_some()
+    }
+}
+
+/// A handle that allows setting a compaction frontier.
+pub trait CompactionHandle {
+    /// Unblocks compaction for updates at or before `since`.
+    fn allow_compaction(&self, since: Antichain<Timestamp>);
+}
+
+/// A handle that allows setting a compaction frontier that is bounded by an upper bound.
+struct BoundedCompactionHandle {
+    /// Returns the upper bound when asked to.
+    upper_bound_fn: Box<dyn Fn() -> Antichain<Timestamp>>,
+
+    /// Underlying compaction handle.
+    compaction_handle: MultiWriteHandle,
+}
+
+impl BoundedCompactionHandle {
+    /// Creates a new [`BoundedCompactionHandle`] that bounds compaction using the given
+    /// `upper_bound_fn` and forwards compaction requests to the given `compaction_handle`
+    pub fn new<F>(upper_bound_fn: F, compaction_handle: MultiWriteHandle) -> Self
+    where
+        F: Fn() -> Antichain<Timestamp> + 'static,
+    {
+        Self {
+            upper_bound_fn: Box::new(upper_bound_fn),
+            compaction_handle,
+        }
+    }
+}
+
+impl CompactionHandle for BoundedCompactionHandle {
+    fn allow_compaction(&self, since: Antichain<Timestamp>) {
+        let mut upper_bound = (*self.upper_bound_fn)();
+
+        println!(
+            "persistence frontier {:?}, allowed_compaction: {:?}",
+            upper_bound, since
+        );
+
+        upper_bound.meet_assign(&since);
+
+        println!("combined compaction frontier: {:?}", upper_bound,);
+
+        // NOTE: We simply shoot off the `allow_compaction()` request and don't even block
+        // on the response. This will mean we don't get any feedback and log errors. But
+        // compactions are not necessary for correctness and there are means (metrics) for figuring
+        // out if the frontier does in fact advance.
+        self.compaction_handle.allow_compaction_all(upper_bound);
```

Comment:
nit: assign this to `let _ = ` to doubly reinforce that we're intentionally throwing this away